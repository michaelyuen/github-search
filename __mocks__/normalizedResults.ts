export const normalizedResults = [
  {
    description:
      '<div><img class="emoji" title=":atom:" alt=":atom:" src="https://github.githubassets.com/images/icons/emoji/atom.png" height="20" width="20" align="absmiddle"> The hackable text editor</div>',
    id: "MDEwOlJlcG9zaXRvcnkzMjI4NTA1",
    license: "MIT license",
    name: "atom",
    owner: "atom",
    stars: 58554,
    url: "https://github.com/atom/atom",
  },
  {
    description:
      "<div>The Atomic Game Engine is a multi-platform 2D and 3D engine with a consistent API in C++, C#, JavaScript, and TypeScript</div>",
    id: "MDEwOlJlcG9zaXRvcnkyOTc5MDcwNw==",
    license: "OTHER license",
    name: "AtomicGameEngine",
    owner: "AtomicGameEngine",
    stars: 2764,
    url: "https://github.com/AtomicGameEngine/AtomicGameEngine",
  },
  {
    description: "<div>Java course materials</div>",
    id: "MDEwOlJlcG9zaXRvcnk2ODY0Mzk3MQ==",
    license: "MIT license",
    name: "atom",
    owner: "rybalkinsd",
    stars: 307,
    url: "https://github.com/rybalkinsd/atom",
  },
  {
    description:
      "<div>Small and highly portable detection tests based on MITRE's ATT&amp;CK.</div>",
    id: "MDEwOlJlcG9zaXRvcnkxMDY1ODYxMjQ=",
    license: "MIT license",
    name: "atomic-red-team",
    owner: "redcanaryco",
    stars: 6584,
    url: "https://github.com/redcanaryco/atomic-red-team",
  },
  {
    description:
      "<div>Atom file-specific icons for improved visual grepping.</div>",
    id: "MDEwOlJlcG9zaXRvcnkxODE1NjIxMA==",
    license: "MIT license",
    name: "atom",
    owner: "file-icons",
    stars: 1238,
    url: "https://github.com/file-icons/atom",
  },
  {
    description:
      "<div>A reactive Java framework for building fault-tolerant distributed systems</div>",
    id: "MDEwOlJlcG9zaXRvcnkxNjI5NDkyMA==",
    license: "APACHE-2.0 license",
    name: "atomix",
    owner: "atomix",
    stars: 2250,
    url: "https://github.com/atomix/atomix",
  },
  {
    description:
      '<div><g-emoji class="g-emoji" alias="book" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/1f4d6.png">üìñ</g-emoji> Documentation for Atom, generated by nanoc, hosted by GitHub Pages</div>',
    id: "MDEwOlJlcG9zaXRvcnkyNTAwMzA3MA==",
    license: "OTHER license",
    name: "flight-manual.atom.io",
    owner: "atom",
    stars: 607,
    url: "https://github.com/atom/flight-manual.atom.io",
  },
  {
    description:
      '<div><g-emoji class="g-emoji" alias="memo" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/1f4dd.png">üìù</g-emoji> Markdown preview in Atom</div>',
    id: "MDEwOlJlcG9zaXRvcnkxMjA4NzU3NQ==",
    license: "MIT license",
    name: "markdown-preview",
    owner: "atom",
    stars: 1235,
    url: "https://github.com/atom/markdown-preview",
  },
  {
    description: "<div>Atom Package Manager</div>",
    id: "MDEwOlJlcG9zaXRvcnk5OTE2Mzc5",
    license: "MIT license",
    name: "apm",
    owner: "atom",
    stars: 1258,
    url: "https://github.com/atom/apm",
  },
  {
    description:
      '<div><g-emoji class="g-emoji" alias="mega" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/1f4e3.png">üì£</g-emoji> Help Wanted - Looking for Maintainer: <a href="https://github.com/Glavin001/atom-beautify/issues/2572">https://github.com/Glavin001/atom-beautify/issues/2572</a> | <g-emoji class="g-emoji" alias="lipstick" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/1f484.png">üíÑ</g-emoji> Universal beautification package for Atom editor (<g-emoji class="g-emoji" alias="warning" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/26a0.png">‚ö†Ô∏è</g-emoji> Currently migrating to <a href="https://github.com/Unibeautify/">https://github.com/Unibeautify/</a> and have very limited bandwidth for Atom-Beautify Issues. Thank you for your patience and understanding <g-emoji class="g-emoji" alias="heart" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/2764.png">‚ù§Ô∏è</g-emoji> )</div>',
    id: "MDEwOlJlcG9zaXRvcnkxNzM1NjkxMQ==",
    license: "MIT license",
    name: "atom-beautify",
    owner: "Glavin001",
    stars: 1507,
    url: "https://github.com/Glavin001/atom-beautify",
  },
];
