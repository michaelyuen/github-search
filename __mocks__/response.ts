export const data = {
  nodes: [
    {
      description: ":atom: The hackable text editor",
      id: "MDEwOlJlcG9zaXRvcnkzMjI4NTA1",
      licenseInfo: {
        name: "MIT License",
        __typename: "License",
      },
      name: "atom",
      stargazerCount: 58550,
      url: "https://github.com/atom/atom",
      __typename: "Repository",
    },
    {
      description:
        "Share your workspace with team members and collaborate on code in real time in Atom",
      id: "MDEwOlJlcG9zaXRvcnk5MzUwMDAzMw==",
      licenseInfo: {
        name: "MIT License",
        __typename: "License",
      },
      name: "teletype",
      stargazerCount: 2383,
      url: "https://github.com/atom/teletype",
      __typename: "Repository",
    },
    {
      description: "Next generation vim support for atom",
      id: "MDEwOlJlcG9zaXRvcnkxMTgyNzEwNA==",
      licenseInfo: {
        name: "MIT License",
        __typename: "License",
      },
      name: "vim-mode",
      stargazerCount: 1719,
      url: "https://github.com/atom/vim-mode",
      __typename: "Repository",
    },
    {
      description: "Native Password Node Module",
      id: "MDEwOlJlcG9zaXRvcnk5NjgyODM0",
      licenseInfo: {
        name: "MIT License",
        __typename: "License",
      },
      name: "node-keytar",
      stargazerCount: 1325,
      url: "https://github.com/atom/node-keytar",
      __typename: "Repository",
    },
    {
      description: "Atom Package Manager",
      id: "MDEwOlJlcG9zaXRvcnk5OTE2Mzc5",
      licenseInfo: {
        name: "MIT License",
        __typename: "License",
      },
      name: "apm",
      stargazerCount: 1258,
      url: "https://github.com/atom/apm",
      __typename: "Repository",
    },
    {
      description: "üìù Markdown preview in Atom",
      id: "MDEwOlJlcG9zaXRvcnkxMjA4NzU3NQ==",
      licenseInfo: {
        name: "MIT License",
        __typename: "License",
      },
      name: "markdown-preview",
      stargazerCount: 1235,
      url: "https://github.com/atom/markdown-preview",
      __typename: "Repository",
    },
    {
      description: ":octocat: Git and GitHub integration for Atom",
      id: "MDEwOlJlcG9zaXRvcnk0NjE1NTc4Ng==",
      licenseInfo: {
        name: "MIT License",
        __typename: "License",
      },
      name: "github",
      stargazerCount: 1048,
      url: "https://github.com/atom/github",
      __typename: "Repository",
    },
    {
      description:
        "View and insert possible completions in the editor while typing",
      id: "MDEwOlJlcG9zaXRvcnkxNzI1NDcyOA==",
      licenseInfo: {
        name: "MIT License",
        __typename: "License",
      },
      name: "autocomplete-plus",
      stargazerCount: 965,
      url: "https://github.com/atom/autocomplete-plus",
      __typename: "Repository",
    },
    {
      description:
        "String-wise sequence CRDT powering peer-to-peer collaborative editing in Teletype for Atom.",
      id: "MDEwOlJlcG9zaXRvcnk5MTU5NzA1Ng==",
      licenseInfo: {
        name: "MIT License",
        __typename: "License",
      },
      name: "teletype-crdt",
      stargazerCount: 726,
      url: "https://github.com/atom/teletype-crdt",
      __typename: "Repository",
    },
    {
      description:
        "üìñ Documentation for Atom, generated by nanoc, hosted by GitHub Pages",
      id: "MDEwOlJlcG9zaXRvcnkyNTAwMzA3MA==",
      licenseInfo: {
        name: "Other",
        __typename: "License",
      },
      name: "flight-manual.atom.io",
      stargazerCount: 608,
      url: "https://github.com/atom/flight-manual.atom.io",
      __typename: "Repository",
    },
  ],
  repositoryCount: 221,
  __typename: "SearchResultItemConnection",
};
