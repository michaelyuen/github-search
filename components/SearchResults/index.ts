export * from "./SearchResult";
export * from "./styles";
export * from "./utils";
