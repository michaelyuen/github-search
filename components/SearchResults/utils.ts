import { SearchResultsProps } from ".";
import { Data } from "../../graphql";

export const normalizeRepos = (
  data: Data
): SearchResultsProps["results"] | null => {
  if (!data?.repos?.nodes) {
    return null;
  }

  return data.repos.nodes.map((repo) => ({
    description: repo.descriptionHTML,
    id: repo.id,
    license: repo.licenseInfo?.name ?? null,
    name: repo.nameWithOwner,
    stars: repo.stargazerCount,
    url: repo.url,
  }));
};
