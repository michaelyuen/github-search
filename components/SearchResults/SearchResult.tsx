import { SearchResultsContainer } from "./styles";
import { SearchResult } from "../SearchResult";

export interface SearchResultsProps {
  noResultsMessage: string;
  results?: SearchResult[];
  title: string;
}

export const SearchResults: React.FC<SearchResultsProps> = ({
  noResultsMessage,
  results,
  title,
}) => {
  const hasResults = results && results.length > 0;
  return (
    <SearchResultsContainer>
      <h3>{title}</h3>
      {!hasResults && <p>{noResultsMessage}</p>}
      {hasResults &&
        results.map((result, i) => (
          <SearchResult key={result.id} {...result} />
        ))}
    </SearchResultsContainer>
  );
};
