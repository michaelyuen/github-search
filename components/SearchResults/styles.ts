import styled from "styled-components";
import { SearchResultContainer } from "../SearchResult/styles";

export const SearchResultsContainer = styled.div`
  ${SearchResultContainer} {
    margin-bottom: 1em;
  }
`;
