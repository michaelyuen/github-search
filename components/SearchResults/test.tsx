import { render, screen } from "@testing-library/react";
import { axe } from "jest-axe";
import { SearchResults } from "./SearchResults";
import { normalizedResults } from "../../__mocks__/normalizedResults";

describe("SearchResults", () => {
  test("renders", () => {
    render(
      <SearchResults
        aria-label="mock aria label"
        noResultsMessage=""
        results={[]}
        title="mock title"
      />
    );
    expect(screen.getByLabelText("mock aria label")).toBeInTheDocument();
  });

  test("matches snapshot", () => {
    const { container } = render(
      <SearchResults
        aria-label="mock aria label"
        noResultsMessage=""
        title="mock title"
      />
    );
    expect(container).toMatchSnapshot();
  });

  test("has no a11y violations", async () => {
    const { container } = render(
      <SearchResults
        aria-label="mock aria label"
        noResultsMessage=""
        title="mock title"
      />
    );
    expect(await axe(container)).toHaveNoViolations();
  });

  test("renders with mock data", () => {
    render(
      <SearchResults
        aria-label="mock aria label"
        noResultsMessage=""
        results={normalizedResults}
        title="mock title"
      />
    );
    expect(screen.getByLabelText("mock aria label")).toBeInTheDocument();
  });
});
