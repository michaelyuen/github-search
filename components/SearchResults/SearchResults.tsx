import { SearchResultsContainer } from "./styles";
import { SearchResult, SearchResultProps } from "../SearchResult";
import { ComponentProps } from "../types";

const defaultMessage = (
  <p>
    Please enter a query above.{" "}
    <a
      href="https://docs.github.com/en/enterprise-cloud@latest/search-github/searching-on-github/searching-for-repositories"
      rel="noopener noreferrer"
      target="_blank"
    >
      See here for syntax.
    </a>
  </p>
);
const defaultNoResultsMessage =
  "No results. Please update your query and try again.";
const loadingItems = Array.from("x".repeat(10));

export interface SearchResultsProps extends ComponentProps {
  initialMessage?: string | JSX.Element;
  hasSearched: boolean;
  isLoading: boolean;
  noResultsMessage?: string | JSX.Element;
  results?: SearchResultProps[];
  title: string;
}

export const SearchResults: React.FC<SearchResultsProps> = ({
  initialMessage = defaultMessage,
  hasSearched,
  isLoading,
  noResultsMessage = defaultNoResultsMessage,
  results,
  title,
  ...props
}) => {
  const hasResults = results && results.length > 0;
  return (
    <SearchResultsContainer {...props}>
      <h2>{title}</h2>
      {isLoading &&
        loadingItems.map((_, i) => (
          // Explicit ignore because `isLoading` will render a skeleton and not use any data
          // @ts-ignore
          <SearchResult isLoading key={i} />
        ))}
      {!isLoading && !hasSearched && <div>{initialMessage}</div>}
      {!isLoading && hasSearched && !hasResults && <p>{noResultsMessage}</p>}
      {!isLoading && hasSearched && hasResults && (
        <ul>
          {results.map((result, i) => (
            <SearchResult key={result.id} {...result} />
          ))}
        </ul>
      )}
    </SearchResultsContainer>
  );
};
