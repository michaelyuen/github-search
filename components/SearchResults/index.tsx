import { SearchResultsContainer } from "./SearchResults.styles";

export interface SearchResultsProps {
  noResultsMessage: string;
  results?: Array<{
    description: string;
    fork: boolean;
    license: string;
    name: string;
    stars: number;
    url: string;
  }>;
  title: string;
}

export const SearchResults: React.FC<SearchResultsProps> = ({
  noResultsMessage,
  results,
  title,
}) => {
  const hasResults = results && results.length > 0;
  return (
    <SearchResultsContainer>
      <h3>{title}</h3>
      {!hasResults && <p>{noResultsMessage}</p>}
      {hasResults && results.map((result, i) => <p key={i}>i am a result</p>)}
    </SearchResultsContainer>
  );
};
export { SearchResultsContainer };
