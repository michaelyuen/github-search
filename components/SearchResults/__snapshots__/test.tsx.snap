// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults matches snapshot 1`] = `
<div>
  <div
    aria-label="mock aria label"
    class="sc-jSUZER jUNxna"
  >
    <h2>
      mock title
    </h2>
    <div>
      <p>
        Please enter a query above. 
        <span>
          This search only supports the “user” and “stars” qualifiers.
        </span>
         
        <a
          href="https://docs.github.com/en/enterprise-cloud@latest/search-github/searching-on-github/searching-for-repositories"
          rel="noopener noreferrer"
          target="_blank"
        >
          See here for syntax.
        </a>
      </p>
    </div>
  </div>
</div>
`;
